src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -349,6 +349,46 @@ def _prepare_for_mds_upgrade(
 
         return continue_upgrade
 
+    def _post_upgrade_redeploys(self):
+        # type: () -> None
+        daemons = self.mgr.cache.get_daemons()
+
+        # redeploy all daemons who are not redployed during upgrade
+        # and mgrs which could have been deployed by a mgr running old code
+        for d in daemons:
+            assert d.daemon_type is not None
+            assert d.daemon_id is not None
+            assert d.hostname is not None
```

Comment:
I really need to fix the types. having those members being optional is irritating.